lib/elixir/unicode/tokenizer.ex

```diff
@@ -112,137 +106,362 @@ defmodule String.Tokenizer do
+      acc ->
+        Map.update!(acc, key, fn value ->
+          additions
+          |> Enum.map(&Map.fetch!(acc, &1))
+          |> Enum.reduce(value, &ScriptSet.union/2)
+        end)
+    end
 
-  range = rangify.(unicode_upper)
+  {bottom, top} = ScriptSet.lattices(map_size(scriptset_masks))
+  IO.puts(:stderr, "[Unicode] Tokenizing #{map_size(scriptset_masks)} scriptsets")
 
-  for {first, last} <- range do
-    if first == last do
-      defp unicode_upper?(unquote(first)), do: true
-    else
-      defp unicode_upper?(entry) when entry in unquote(first)..unquote(last), do: true
+  codepoints_to_mask =
+    for {codepoint, scriptsets} <- all_codepoints_to_scriptset, into: %{} do
+      {codepoint,
+       scriptsets
+       |> Enum.map(&Map.fetch!(scriptset_masks, &1))
+       |> Enum.reduce(bottom, &ScriptSet.union/2)}
     end
+
+  ##
+  ## Define functions and module attributes to access characters and their scriptsets
+  ##
+
+  @bottom bottom
+  @latin 1
+  @top top
+  @indexed_scriptsets sorted_scriptsets |> Enum.with_index(&{&2, &1}) |> Map.new()
+
+  latin = Map.fetch!(scriptset_masks, "Latin")
+
+  @highly_restrictive [
+    ScriptSet.union(latin, Map.fetch!(scriptset_masks, "Japanese")),
+    ScriptSet.union(latin, Map.fetch!(scriptset_masks, "Han with Bopomofo")),
+    ScriptSet.union(latin, Map.fetch!(scriptset_masks, "Korean"))
+  ]
+
+  # ScriptSet helpers. Inline instead of dispatching to ScriptSet for performance
+
+  @compile {:inline, ss_latin: 1, ss_intersect: 2}
+  defp ss_latin(ss), do: :erlang.band(ss, @latin)
+  defp ss_intersect(left, right), do: :erlang.band(left, right)
+
+  # Ascii helpers
+
+  @compile {:inline, ascii_upper?: 1, ascii_lower?: 1, ascii_continue?: 1}
+  defp ascii_upper?(entry), do: entry >= ?A and entry <= ?Z
+  defp ascii_lower?(entry), do: entry >= ?a and entry <= ?z
+  defp ascii_continue?(entry), do: entry >= ?0 and entry <= ?9
+
+  # Pattern is used as a performance check to end sooner before traversing unicode
+  for pattern <- ' \t\n\r!"#$%&\'()*+,-./:;<=>?@[]^`{|}~' do
+    defp ascii_pattern?(unquote(pattern)), do: true
   end
 
-  defp unicode_upper?(_), do: false
+  defp ascii_pattern?(_), do: false
+
+  # Unicode helpers
+  # We use ranges whenever possible to reduce bytecode size.
+
+  unicode_upper = Enum.map(unicode_upper, &{&1, Map.get(codepoints_to_mask, &1, top)})
+  unicode_start = Enum.map(unicode_start, &{&1, Map.get(codepoints_to_mask, &1, top)})
+  unicode_continue = Enum.map(unicode_continue, &{&1, Map.get(codepoints_to_mask, &1, top)})
+
+  rangify = fn [{head, scriptset} | tail] ->
+    {first, last, scriptset, acc} =
+      Enum.reduce(tail, {head, head, scriptset, []}, fn
+        {number, scriptset}, {first, last, scriptset, acc} when number == first - 1 ->
+          {number, last, scriptset, acc}
+
+        {number, scriptset}, {first, last, range_scriptset, acc} ->
+          {number, number, scriptset, [{first, last, range_scriptset} | acc]}
```

Comment:
On a first read, I worried that there was a risk this wouldn't work at boundaries where one range is in one scriptset, and the other is in another scriptset. But looking closer, I see I was overlooking that `scriptset` is matched with the accumulated/current one in that clause 👍